tapdb/universe_federation_test.go

```diff
@@ -98,6 +87,232 @@ func TestUniverseFederationCRUD(t *testing.T) {
+		universe.SyncDirectionPush, universe.ProofSyncStatusPending,
+		false,
+	)
+	require.NoError(t, err)
+
+	// Check that the attempt counter was not changed as expected.
+	pendingLogEntries, err = fedStore.QueryFederationProofSyncLog(
+		ctx, uniId, uniProof.LeafKey, universe.SyncDirectionPush,
+		universe.ProofSyncStatusPending,
+	)
+	require.NoError(t, err)
+	require.Len(t, pendingLogEntries, 2)
+
+	for i := range pendingLogEntries {
+		entry := pendingLogEntries[i]
+		if entry.ServerAddr == servers[1] {
+			require.Equal(t, int64(1), entry.AttemptCounter)
+		} else {
+			require.Equal(t, int64(0), entry.AttemptCounter)
+		}
+	}
+
+	// Set the sync status to complete for one of the pending log entries.
+	_, err = fedStore.UpsertFederationProofSyncLog(
+		ctx, uniId, uniProof.LeafKey, servers[1],
+		universe.SyncDirectionPush, universe.ProofSyncStatusComplete,
+		false,
+	)
+	require.NoError(t, err)
+
+	// Check that the sync status was updated as expected.
+	pendingLogEntries, err = fedStore.QueryFederationProofSyncLog(
+		ctx, uniId, uniProof.LeafKey, universe.SyncDirectionPush,
+		universe.ProofSyncStatusPending,
+	)
+	require.NoError(t, err)
+	require.Len(t, pendingLogEntries, 1)
+
+	completeLogEntries, err = fedStore.QueryFederationProofSyncLog(
+		ctx, uniId, uniProof.LeafKey, universe.SyncDirectionPush,
+		universe.ProofSyncStatusComplete,
+	)
+	require.NoError(t, err)
+	require.Len(t, completeLogEntries, 2)
+
+	// Delete log entries for one of the servers.
+	err = fedStore.DeleteProofsSyncLogEntries(ctx, servers[0], servers[1])
+	require.NoError(t, err)
+
+	// Only one log entry should remain and it should have sync status
+	// pending.
+	pendingLogEntries, err = fedStore.QueryFederationProofSyncLog(
+		ctx, uniId, uniProof.LeafKey, universe.SyncDirectionPush,
+		universe.ProofSyncStatusPending,
+	)
+	require.NoError(t, err)
+	require.Len(t, pendingLogEntries, 1)
+
+	// Check that the remaining log entry is as expected.
+	pendingEntry := pendingLogEntries[0]
+	require.Equal(t, servers[2], pendingEntry.ServerAddr)
+}
+
+// assertProofSyncLogLeafKey asserts that a leaf key derived from a proof sync
+// log entry is equal to a given leaf key.
+func assertProofSyncLogLeafKey(t *testing.T, actualLeafKey universe.LeafKey,
+	logLeafKey universe.LeafKey) {
+
+	// We can safely ignore the tweaked script key as it is the derivation
+	// information for the script key. It is only ever known to the owner of
+	// the asset and is never serialized in a proof
+	actualLeafKey.ScriptKey.TweakedScriptKey = nil
+	require.Equal(t, actualLeafKey, logLeafKey)
+}
+
+// assertProofSyncLogLeaf asserts that a leaf derived from a proof sync log
+// entry is equal to a given universe leaf.
+func assertProofSyncLogLeaf(t *testing.T, actualLeaf universe.Leaf,
+	logLeaf universe.Leaf) {
+
+	if actualLeaf.GenesisWithGroup.GroupKey != nil {
+		actualLeaf.GenesisWithGroup.GroupKey.Witness = nil
+		actualLeaf.GenesisWithGroup.GroupKey.RawKey.PubKey = nil
```

Comment:
the raw pubkey here is also populated via some other means I assume?

Reply:
I don't think so. In the leaf derived from the log, the `RawKey.PubKey` field is not set.

I think that's OK because the distributed leaf doesn't need to include any data that was used to derive the group key. As far as I understand, the the `RawKey` is the pre-teaked group key. Do you agree?